### ISSUE TITLE
Handling different types of stations

### ISSUE THREAD

**Issue by maxcal** [remote-wind/remote-wind#161]
We need to add some way to differentiate between Ardiuno and Node stations.
I'm thinking of doing this as Single Table Inheritance and setting up a class tree:

```
class Station
end

# this is nested in Stations  due to the way ActiveRecord loads namespaced classes
module Stations
  class Ardiuno < ::Station
  end
end

module Stations
  class NodeRed < ::Station
  end
end
```

**Comment by maxcal** [comment]
This lets us set up callbacks and handle stuff like clearing out observations in a clean way without creating a ton of code paths.

**Comment by karl-petter** [comment]
Make it more generic and focus on functionality rather than the source of the data.

The current firmware of the arduinos does not support pre-paid card balance reporting but might be something to add back in the future.

**Comment by maxcal** [comment]
We could add feature flags but doing a major split between the ardiuno type and the other stations that don't need all the utility stuff like reporting, the clunky saving raw data and calibrating it at runtime etc should be done

**Comment by maxcal** [comment]
I'm not to hot on the idea of adding any more config flags than strictly neccisary.

**Comment by maxcal** [comment]
You have to look at the fact that a lot of stuff like callbacks, validations etc are happing on the class level. While you can add a bunch of instance level feature flags and a bunch of conditions on everything it will just further bloat the station class and be more difficult to test.

**Comment by karl-petter** [comment]
I do not agree. Both kinds are stations that report wind information and share much more than they do not share, so why then should it be split into two classes? For instance the reporting is needed for the any kind of station since without input of data a station is broken.

**Comment by karl-petter** [comment]
I have created a new version of the firmware that integrates the scaling into the Arduino firmware so if the scaling is correct the values from station with ID 22 should be "true" but the problem is if it turns out that the factor is wrong it is but hard to change since a laptop is needed to brought to the station and a new firmware be downloaded onto the Arduino board. Thus I would strongly argue for keeping a scaling factor. And I think it even can make sense for none Arduino stations, for instance we know that the Skistar stations are not totally reliable and geographic differences makes stations to measure quite differently.

**Comment by maxcal** [comment]
Yeah but we don't need to save raw data and convert the records when they are loaded back which uses a bunch of memory on every non cached get request.

**Comment by karl-petter** [comment]
Are we running low on RAM on the server?

**Comment by maxcal** [comment]
No, but we only have one dyno. Anything that ties up the dyno like looping through all the fetched observerations is better avoided.

**Comment by karl-petter** [comment]
We could always send both values to the frontend which can make the calculation instead if it is an issue. Do we have any numbers indicating that it is a problem?

But I still do not get what the problem is with memory. Could you please explain it more?

**Comment by maxcal** [comment]
Basically we have a single web dyno which serves up all the requests - whenever it processes a request to `/stations/:station_id/observations` it both has to do a slow database query to the observations table and then has to loop through every observation fetched and multiple the columns. The main issue is not really memory but rather execution time. 

I belive we have the wrong priority - the main point of the app is not scientific data or record keeping. At least the reason I'm working on it is I want to be able to be able to when the wind is good for kitesurfing. 

All these kind of eccentric requirements take a bunch of time off things that really matter like revamping the frontend and changing the mapping layer. So what if you recalibrate a station and the old records may be off sync? 

If we want to run on bottom tier hosting we need to prioritize simplicity and performance over features with marginal benefits.

**Comment by maxcal** [comment]
Doing it on the frontend will be problematic for weak devices and lead to a duplication of the logic all over this place since the observation data is rendered server side in quite a few places. Its also a horrible architecture descision. Serving up accurate data is the servers job.

**Comment by karl-petter** [comment]
The main problem is that the calibration has to be made on spot. I'm not very fond of bringing my laptop to the top of Rödkullen in the middle of the winter, take down the mast, plugin the USB, flash the firmware, verify that all works again and then get back for just changing a number. But if you can take on that responsibility to change calibration when needed, please feel free to remove it from the server.

And regarding priorities, when I surf the site it feels really responsive so why this focus to improve speed when it does not seem to be a problem? What kind of time savings are we talking about? What is a typical request timing today? How low can we get it?

**Comment by maxcal** [comment]
The way I designed it feels responsive since the HTML routes actually do very little and most of the actual data is retrieved in an ajax request. However fetching `/stations.json` takes 3-10s which is slow.

**Comment by karl-petter** [comment]
And how long time does it take to fetch the stations if no conversion is made of the observations?

**Comment by maxcal** [comment]
I would need to do clone of the pg db to test that. Won't have time until later this week
